### ISSUE TITLE
Black-related Errors

### ISSUE THREAD

**Issue by gnatenko** [iamcrowd/t-crowd-lib#10]
Example:

error: cannot format FchdV.pltl: Cannot parse: 1:19: ( ((Rttgm_Adult_t1 =>  ~ Rttgm_Minor_t1) & (Rttgm_Adult_t1 => Rttgm_Person_t1) & (Rttgm_Minor_t1 => Rttgm_Person_t1) & Rttgm_Adult_t0 & Rttgm_Minor_t1))

Can it be due to spaces after the first bracket?

**Comment by nicola-gigante** [comment]
Hi guys. This does not seem an error generated by BLACK. How do you call it?

**Comment by gnatenko** [comment]
Basically, we invoke a command-line instance of Black, but from java code. Like this:

        ProcessBuilder pb = new ProcessBuilder();
        pb.command("black", file2);
        pb.redirectErrorStream(true);

        Process p5 = pb.start();

        StringBuilder output = new StringBuilder();
        BufferedReader reader = new BufferedReader(
                                new InputStreamReader(p5.getInputStream()));

        String line;
        while ((line = reader.readLine()) != null) {
            output.append(line + "\n");
        }

        int exitVal = p5.waitFor();

So, "output" should contain the output of Black. And it contains what I've posted.

NB: German and I use different versions of Black. I installed mine on Mac using Homebrew, and it is (supposedly) an older Version, that doesn't use "solve" keyword.

I learned not to use "solve" also form "output", where Black told me "solve is not a path" or something like that. So, chances are "output" is real Black's output.

There is also "Oh no!" and a broken heart emoji in that "output"... Does it seem possible?

**Comment by gnatenko** [comment]
Black gives the same message when I feed the pltl file to it in the command line. So it seems to be really Black's output. But maybe from an older Version on Mac?

**Comment by nicola-gigante** [comment]
Ahah, no, BLACK does not print emojis, but that might be a good idea for future versions 🤡 The homebrew package is always up to date to the latest version, but you installed the wrong package. If you do `brew install black`, it installs "another" black, which is a code formatter for python (which explains the error "cannot format"). The name of the package was already taken! :)

"Our" black is on a separate tap, the installation instructions are on GitHub:

```
$ brew tap black-sat/black
$ brew install black-sat
```

You probably have to uninstall the other package before that.

**Comment by gnatenko** [comment]
Wow! A python code formatter... That could have been a very interesting DL-Lite tool! :D
I thought I took the commands from Github, but I will redo the installation. Thanks!

**Comment by gnatenko** [comment]
Still not working :/

So, I installed the right Black. Homebrew puts it to /opt/homebrew/Cellar/black-sat/0.8.0/. It seems tricky to uninstall the other black though, the code formatter. So I went directly to the directory and tried to invoke Black:


```
bin $ ./black
dyld[82173]: Library not loaded: @rpath/libblack.dylib
  Referenced from: /opt/homebrew/Cellar/black-sat/0.8.0/bin/black
  Reason: tried: '/usr/local/lib/libblack.dylib' (no such file), '/usr/lib/libblack.dylib' (no such file)
zsh: abort      ./black
```

I tried adding /opt/homebrew/Cellar/black-sat/0.8.0/lib/libblack.dylib to LD_LIBRARY_PATH, but it didn't help.

Maybe you have some quick ideas of what's wrong? Otherwise we will have to run tests on Herman's virtual machine, I suppose...

**Comment by gnatenko** [comment]
Ok, I somehow made my Mac to copy the lib to usr/local/lib. Hope it wouldn't waive the warranty... :D

**Comment by nicola-gigante** [comment]
I'm sorry, that's a problem we currently have with the homebrew package on M1 Macs. I just discovered it some days ago because I didn't have an M1 before that. I still have to understand how to fix it. So did you manage to make it work by manually copying the library?

**Comment by gnatenko** [comment]
Yep, now it works. Although there are still some issues with our library, but at least Black works

**Comment by gnatenko** [comment]
On some files (my version of) Black gives segmentation fault. I attach the file just in case (it's quite big). The formula was generated by the lib, and it may well be broken, but still. Maybe it's also an M1 issue? Could someone try to run it elsewhere? The file [f.pltl.zip](https://github.com/iamcrowd/t-crowd-lib/files/8934372/f.pltl.zip) is attached (as a zip, since GitHub doesn't like .pltl)


`
$ /opt/homebrew/Cellar/black-sat/0.8.0/bin/black solve f.pltl

zsh: segmentation fault  /opt/homebrew/Cellar/black-sat/0.8.0/bin/black solve f.pltl
`

P.S. Sorry for late Saturday night emails, ignore them if you have better things to do :)

**Comment by nicola-gigante** [comment]
Hi Anton, I've tried that formula on a Linux server and it crashed as well, so unfortunately it's not a problem of the M1 Macs. I'll try to understand what is going on.

**Comment by nicola-gigante** [comment]
Hi, luckily the fix was easy. The updated version is already online. If you do not want to have problems with the Homebrew package which still needs to be fixed, I suggest you to build from source. It should not be hard, Homebrew already installed all the dependencies for you. Just follow the instructions [here](https://github.com/black-sat/black/wiki/Installation#installation-from-source). Let me know if you have problems, I'm available for a call today if needed.

**Comment by nicola-gigante** [comment]
Hi @gnatenko, I've fixed the problem of the Homebrew package on M1 Macs. Now the installation should be easy again, just `brew install black-sat`, or, in your case `brew upgrade black-sat`. Just, before, remove the library you copied in `/usr/local/lib`, just to avoid potential problems.

To uninstall the other black, does `brew uninstall black` not work?

**Comment by gnatenko** [comment]
Hi Nicola!

Thanks a lot for your help! Now I get segmentation fault on another file :(

[g.pltl.zip](https://github.com/iamcrowd/t-crowd-lib/files/8935150/g.pltl.zip)

But the issue with the lib in homebrew is resolved :)

> To uninstall the other black, does brew uninstall black not work?

It seems that "the other black" was there before I installed homebrew itself. So it needs to be uninstalled some other way. And I haven't activated sudo on my Mac yet, so for the time being I just decided not to care.
